lib/rules/prefer-object-spread.js

```diff
@@ -0,0 +1,278 @@
+/**
+ * @fileoverview Prefers object spread property over Object.assign
+ * @author Sharmila Jesupaul
+ * See LICENSE file in root directory for full license.
+ */
+
+"use strict";
+
+const matchAll = require("string.prototype.matchall");
+
+/**
+ * Helper that checks if the node is an Object.assign call
+ * @param {ASTNode} node - The node that the rule warns on
+ * @returns {boolean} - Returns true if the node is an Object.assign call
+ */
+function isObjectAssign(node) {
+    return (
+        node.callee &&
+        node.callee.type === "MemberExpression" &&
+        node.callee.object.name === "Object" &&
+        node.callee.property.name === "assign"
+    );
+}
+
+/**
+ * Helper that checks if the node needs parentheses to be valid JS.
+ * The default is to wrap the node in parentheses to avoid parsing errors.
+ * @param {ASTNode} node - The node that the rule warns on
+ * @returns {boolean} - Returns true if the node needs parentheses
+ */
+function needsParens(node) {
+    const parent = node.parent;
+
+    if (!parent || !node.type) {
+        return true;
+    }
+
+    switch (parent.type) {
+        case "VariableDeclarator":
+        case "ArrayExpression":
+        case "ReturnStatement":
+        case "CallExpression":
+        case "Property":
+            return false;
+        default:
+            return true;
+    }
+}
+
+/**
+ * Determines if an argument needs parentheses. The default is to not add parens.
+ * @param {ASTNode} node - The node to be checked.
+ * @returns {boolean} True if the node needs parentheses
+ */
+function argNeedsParens(node) {
+    if (!node.type) {
+        return false;
+    }
+
+    switch (node.type) {
+        case "AssignmentExpression":
+        case "ArrowFunctionExpression":
+        case "ConditionalExpression":
+            return true;
+        default:
+            return false;
+    }
+}
+
+/**
+ * Helper that adds a comma after the last non-whitespace character that is not a part of a comment.
+ * @param {string} formattedArg - String of argument text
+ * @param {array} comments - comments inside the argument
+ * @returns {string} - argument with comma at the end of it
+ */
+function addComma(formattedArg, comments) {
+    const nonWhitespaceCharacterRegex = /[^\s\\]/g;
+    const nonWhitespaceCharacters = Array.from(matchAll(formattedArg, nonWhitespaceCharacterRegex));
```

Comment:
Is it necessary to use `matchAll` here? Since this regex doesn't have any capturing groups, it seems like this could just be done with:

```js
const nonWhitespaceCharacters = formattedArg.match(nonWhitespaceCharacterRegex);
```

That would avoid the need to add a dependency on `string.prototype.matchall`.

Reply:
With a global regex, `.match` does not return an array of match objects, so there's no way to get the index (position) of each match - this is the entire reason `.matchAll` is needed.

So, no, I don't think there's any way to avoid the dependency (without inlining it, of course, which would be a terrible idea)